### ISSUE TITLE
Codegen: add `into_owned` for messages with Cow fields

### ISSUE THREAD

**Issue by koivunej** [tafia/quick-protobuf#53]
Generating these will be much having users to write such:

```
fn into_owned(self) -> TypeName<'static> {
    TypeName {
        plain_cow: Cow::Owned(self.plain_cow.to_owned),
        opt_cow: self.opt_cow.map(Cow::to_owned).map(Cow::Owned),
        copy_field: self.copy_field.
        message_field: self.message_field.into_owned()
     }
}
```

**Comment by tafia** [comment]
We could add something like this as a `Config` option (I'd prefer the default code not to have it personally).

I don't get the `static` lifetime constraint. We do not need `Cow` at all in the returned struct. I may be missing something.

**Comment by koivunej** [comment]
> We could add something like this as a Config option (I'd prefer the default code not to have it personally).

Agreed on the option. I assume it does not benefit every use case and keeping the code bloat at minimum makes sense.

> I don't get the static lifetime constraint. We do not need Cow at all in the returned struct.

I've struggled to understand `'static` in this context as well but finally the good people of `#rust-beginners` explained that it means "valid for the lifetime of the application" which can be translated to "it  owns all of it's data." `Cow` allows the generated structs to support both use cases: 

 * lifetime bound to the buffer lifetime
 * owned

Adding `into_owned(self)` does not diminish the value of `Cow` fields. They are very useful if you are using the generated structs as a temporary view into an existing buffer and then perhaps map into them into other structs or enums that are owned (if they need to live longer). I'm not an expert `tokio-rs` programmer, but it's `EasyBuf` abstraction might even allow using the generated structs for longer term storage at the cost of requiring more careful buffer management.

**Comment by koivunej** [comment]
I've been thinking this and after way too many implementations of `fn into_owned(self)` I've been thinking about writing a custom derive for generating these methods. That would remove the need to involve any traits with the messages.

I am currently looking into how this could be implemented. It would have to be in a separate crate.

Procedural macros might also be considered for implementing `MessageWrite::get_size`. The generated method contents (a single long expression) causes issues with rustfmt, even when the generated file is being ignored (last I checked).

**Comment by koivunej** [comment]
This seems implementable as I just pushed: [derive-into-owned](https://github.com/koivunej/derive-into-owned).

Perhaps once the #59 will get merged I'll start to work on an option to inject these `#[derive(IntoOwned)]` into generated structs. The dependency would have to be added by user to the using projects dependencies so probably better to opt-in?

**Comment by tafia** [comment]
> Procedural macros might also be considered for implementing MessageWrite::get_size

I believe that eventually all codegen could be replaced by a custom derive. That said, I am no expert and could be wrong.

I agree that the `into_owned` specifically is really welcomed. I am not sure on the other hand if we need to depend on your `derive-into-owned` crate or if we can simply generate the function directly ... and wait for a full `#[derive(Protobuf)]` which would handle both all the current generated items plus the `into_owned` function.

Thanks for all this work!

**Comment by koivunej** [comment]
> I believe that eventually all codegen could be replaced by a custom derive. That said, I am no expert and could be wrong.

Well actually... I guess you could have:

```
#[derive(Protobuf)]
#[protobuf(descriptor="somefile.pb", generate_into_owned = true)]
pub mod client_messages;
```

This could actually be implemented and pretty easily. Since there does not seem to be an issue about this feature, perhaps we can discuss it here. Roadmap could be:

 1. refactor to use [`quote! { ... }`](https://github.com/dtolnay/quote) macro to generate the code in `pb_rs` 
    * basically String types would change to `quote::Tokens`
 2. wrap the `pb_rs` in a procedural macro
 3. publish some `quick-protobuf-derive` crate

I have not checked how in all ways does #59 change the code generation, but what I remember from the original version I pushed code for, this refactoring will be quite simple. See my `derive-into-owned` for example.
